src/utils/calculateBasicAuthUsernamePasswordPair.js

```diff
@@ -0,0 +1,32 @@
+import { oauth2 } from './oauth2'
+
+export function calculateBasicAuthUsernamePasswordPair ({
+  username,
+  password,
+  token,
+  oauth2format
+}) {
+  const key = [!!username, !!password, !!token, !!oauth2format]
+    .map(Number)
+    .join('')
+  // See helpful truth table in documentation
+  // prettier-ignore
+  switch (key) {
+    case '0000': return null
+    case '1000': throw new Error(`Missing password or token`)
+    case '0100': throw new Error(`Missing username`)
+    case '1100': return { username, password }
+    case '0010': return { username: token, password: '' }
+    case '1010': return { username, password: token }
+    case '0110': throw new Error(`Missing username`)
+    case '1110': throw new Error(`Can't have both password and token`)
+    case '0001': throw new Error(`Missing token`)
+    case '1001': throw new Error(`Missing token`)
+    case '0101': throw new Error(`Missing token`)
+    case '1101': throw new Error(`Missing token`)
+    case '0011': return oauth2(oauth2format, token)
```

Comment:
I think the only valid `??11` state should be `0011`, others ignore `username` and/or `password` even though they are in and that to me indicates the user either made a mistake or expects them to influence the Oath2 flow whereas only token ends up being used and it's not obvious that we'd ignore the user name and password silently when token is passed.

Reply:
Yeah... I guess for some reason I thought users might want to support both in the same call, and if an OAuth2 token is available let that take precedence. But now that I'm thinking about it... they'd probably be very different code paths, because obtaining an OAuth2 is no trivial task. So that scenario where it's convenient probably doesn't exist. Probably wiser to throw errors rather than silently ignore.

Actually, that might just be a good blanket policy: throw an error when combinations of arguments don't make sense, rather than silently ignore arguments. Might be annoying to users, but it also guarantees they are aware of the valid combination.